core/src/test/java/org/jdbi/v3/core/statement/TestQueries.java

```diff
@@ -468,4 +472,58 @@ public void testBindPrimitiveWithExplicitNull() {
                 .one())
             .isNull();
     }
+
+    @Test
+    public void testForEach() {
+        Handle h = h2Extension.openHandle();
+        for (int id = 1; id <= 42; id++) {
+            h.execute("insert into something (id, name) values (?, ?)", id, null);
+        }
+
+        ResultIterable<Integer> ri = h.createQuery("select id from something").mapTo(Integer.class);
+
+        assertThatThrownBy(() -> ri.forEach(null)).isInstanceOf(NullPointerException.class)
+                .hasMessage("Action required");
+
+        AtomicInteger sum = new AtomicInteger(0);
+        int count = ri.forEachWithCount(sum::addAndGet);
+        assertThat(count).isEqualTo(42);
+        assertThat(sum.intValue()).isEqualTo(903);
+    }
+
+    @Test
+    public void testFilter() {
+        Handle h = h2Extension.openHandle();
+
+        int id = 0;
+        for (String name : new String[] {null, "john", "lennon"}) {
+            h.execute("insert into something (id, name) values (?, ?)", ++id, name);
+        }
+
+        ResultIterable<String> ri1 = h.createQuery("select name from something").mapTo(String.class);
+
+        assertThatThrownBy(() -> ri1.filter(null)).isInstanceOf(NullPointerException.class)
+                .hasMessage("Filter required");
+
+        ResultIterable<String> ri2 = ri1
+                .filter(Objects::nonNull)
+                .filter(s -> s.startsWith("j"))
+                .filter(s -> s.contains("h"));
+
+        assertThat(ri1).isNotSameAs(ri2);
+
+        List<String> results = new ArrayList<>();
+
+        ResultIterator<String> iter = ri2.iterator();
+        while (iter.hasNext()) {
+            results.add(iter.next());
+        }
+
+        assertThatThrownBy(iter::next).isInstanceOf(NoSuchElementException.class)
+                .hasMessage("No more filtered results");
+
+        assertThat(results).startsWith("john")
```

Comment:
`assertThat(results).containsExactly("john")` is a more concise way of saying the same thing, I think.

Reply:
very true 🤭